models/base/document.py

```diff
@@ -0,0 +1,231 @@
+from __future__ import annotations
+
+import os
+import pickle
+from typing import Union
+
+from models.base.field import ReferenceDocumentsField
+from models.base.meta_document import MetaDocument, MetaIndexedDocument
+
+
+def persist(func):
+    """
+    Decorator for saving the document after the function call.
+    """
+
+    def wrapper_persist(*args, **kwargs):
+        instance: Document = args[0]
+        func(*args, **kwargs)
+        instance.save()
+
+    return wrapper_persist
+
+
+class Document(metaclass=MetaDocument):
+    """
+    Base class for all documents.
+    Directly subclass this class only for embedded documents.
+    For root level documents (i.e. tables), use the IndexedDocument class.
+    This class is only persisted when they are referenced by IndexedDocument instances.
+
+    To define a document, subclass this class and define attributes as Fields.
+    For example:
+
+        class User(Document):
+            username = Field(primary_key=True)
+            password_hash = Field()
+
+    The attributes are defined with the Field class so that helper function can analyze them.
+    For example, a helper function verifies that only one primary key is defined.
+
+    The __init__ method automatically set field names and values.
+    for example, the above User class can be directly instantiated with the following:
+
+        user = User(username='admin', password_hash='1321A7FE01')
+
+    To customize class initialization, override the __init__ method. For example:
+
+        def __init__(self, username, password):
+            self.validate_username(username)
+            password_hash = self.hash_password(password)
+            super().__init__(username=username, password_hash=password_hash)
+
+    After updating the document, call the save method to persist the change. For example:
+
+        user.username = 'admin2'
+        user.save()
+
+    To delete the document, use the delete method. For example:
+
+        user.delete()
+
+    Deleting a document also automatically removes all references to it.
+    """
+    _initialised = False
+
+    class PrimaryKeyNotDefinedError(Exception):
+        def __init__(self, document):
+            super().__init__(f"Primary key not defined for {document.__class__.__name__}")
+
+    class PrimaryKeyNotSetError(Exception):
+        def __init__(self, field_name):
+            super().__init__(f"Primary key {field_name} not set")
+
+    class ReferrerNotFound(Exception):
+        def __init__(self, referrer_type=None, attribute_name=None):
+            super().__init__(f"Instanced is not referenced by "
```

Comment:
Missing bracket and typo. I think this is meant to read:
```
super().__init__(f"Instance is not referenced by ")
```

Reply:
I also realise that it's missing reference to the f string it's meant to pick up. I'm not sure whether this is meant to be Guessing it should be {referee} but maybe {field_name}?

Reply:
No, wait, never mind - I missed the line after. All is well - resolving.